### ISSUE TITLE
Optimise RAM consuming in Chrome with XHR

### ISSUE THREAD

**Issue by AlttiRi** [Tampermonkey/tampermonkey#1516]
### TL;DR
- There is a memory leak with `responseType: "stream"`.
- With `responseType: "blob"` it consumes a lot of RAM, while it should not at all (in comparison with VM).

### Info

I use Chromium 101 with Tampermonkey BETA 4.17.6161.

I have a script that fetches 500 MB `Blob` to download it.
I host the file locally in main OS, so it downloads fast, I do tests in a VM (Virtual Machine) to explude memory consuming by the server (I use [live-server](https://www.npmjs.com/package/live-server)).

---

### Demo

Here is the demo: [AlttiRi/gm_fetch/demo.user.js](https://github.com/AlttiRi/gm_fetch/blob/master/demo.user.js)
It's big, but the main code is only: [demo.user.js#L221-L245](https://github.com/AlttiRi/gm_fetch/blob/0640bbcae4413e899adb77afaf3f30597277e1d4/demo.user.js#L221-L245)

Install it and open this link:
https://example.com/gm_fetch-demo?url=http://192.168.1.33:8080/1.webm&only-main-demo=1&no-dl=1
- `only-main-demo=1` — to show only 1 demo button
- `no-dl=1` — don't save file on the disk
- `url=http://192.168.1.33:8080/1.webm` a link to your file

As `GM_fetch` it uses this script: https://github.com/AlttiRi/gm_fetch/blob/master/index.js
It's just extended version that I posted before here: https://github.com/Tampermonkey/tampermonkey/issues/1278#issuecomment-1004568936

With `useStream` checkbox it uses `responseType: "stream"` if it supported (TM Beta supports it), w/o — `responseType: "blob"`.

---

#### ViolentMonkey

Lets start with ViolentMonkey BETA v2.13.0.19. 
It does not support `responseType: "stream"`, so ignore the `useStream` checkbox.

VM does not use additional memory (RAM) for big file downloading at all. 
However, it **3 times** write the data on hard drive in `...\Chrome\User Data\Default\blob_storage\...`. 
1500 MB of a temporal data writing to download a 500 MB file.
I assume is uses 500 MB to store Blob in background script, then 500 when it send the data to the content script, and 500 in web page context.

#### TamperMonkey (`"blob"`)

Let's check how TM works. Disable the `useStream` checkbox.
It also write x3 data to hard drive (to blob cache storage), however, it consumes a lot of memory additionally:
![image](https://user-images.githubusercontent.com/16310547/171369215-0f38bffb-eccd-46e6-9b43-c67b996e7a6f.png)

#### TamperMonkey (`"stream"`)

Finally, lets check it with stream approach. Use `useStream` checkbox.

It also wtires tothe cache folder for blob data, but only 500 BM. 3 times less! _(However, sometimes it does not work so fine, and Blob 5MB chunks are also written to Blob cache folder)_
Also there are no peaks on the RAM consuming chart:
![image](https://user-images.githubusercontent.com/16310547/171369473-6ee9e0f5-7f27-4d4c-a18a-e034a4b0d78e.png)

---

In both cases (with streaming and w/o) the garbage collect works.
However, with streaming there is a minor memory leak. It's the problem not with my script, but with the extension.

Run downloading multiple times (with streaming) and close the tab. The memory will not down to the initial value.
...then disable the extension and memory will free:

![image](https://user-images.githubusercontent.com/16310547/171371654-078a94eb-4d9a-4d76-b1b8-0846cc86d7f3.png)

**Comment by AlttiRi** [comment]
### Additional bug (Firefox)

In Firefox with `responseType: "stream"` TM works **very slow** since chunks from background script are sent with notable delay. There is a delay between each chuck sending, that results to taking a lot of time for sending the response entirely.

**Comment by derjanb** [comment]
You can also force a garbage collection. Just open Tampermonkey's background page and select the performance tab.

![image](https://user-images.githubusercontent.com/767504/171399109-bca14420-da6c-4afb-9ca9-7aade6896b08.png)
https://twitter.com/chromedevtools/status/967833574639419392

**Comment by TomLucidor** [comment]
Is this still a "thing"

**Comment by derjanb** [comment]
Should be fixed at 5.5.6233 ([crx](https://tampermonkey.net/crx/tampermonkey_beta_current.crx)|xpi in review)

Chrome/Edge users, please update or install Tampermonkey BETA.
Firefox users, please install the [BETA version](https://www.tampermonkey.net/index.php?browser=firefox) or check for BETA version updates at `about:addons`.

For a quick fix please export your settings and scripts as zip or (JSON) file at the "Utilities" tab and import it back at the fixed BETA version.
